sklearn/linear_model/sag_fast.pyx

```diff
@@ -97,93 +300,139 @@ def sag(SequentialDataset dataset,
+
+                        for class_ind in range(n_classes):
+                            # idx = feature_ind * n_classes + class_ind
+                            idx += 1
+                            weights[idx] -= cum_sum * sum_gradient[idx]
+
+                            # check to see that the weight is not inf or NaN
+                            if not skl_isfinite(weights[idx]):
+                                with gil:
+                                    raise_infinite_error(n_iter)
+
+                        feature_hist[feature_ind] = sample_itr
+
+                # find the current prediction
+                predict_sample(x_data_ptr, x_ind_ptr, xnnz, weights, wscale,
+                               intercept, prediction, n_classes)
+
+                # compute the gradient for this sample, given the prediction
+                if multinomial:
+                    multiloss._dloss(prediction, y, n_classes, sample_weight,
+                                     gradient)
+                else:
+                    gradient[0] = loss._dloss(prediction[0], y) * sample_weight
 
                 # make the updates to the sum of gradients
-                gradient_change = gradient - gradient_memory[current_index]
                 for j in range(xnnz):
-                    idx = x_ind_ptr[j]
+                    feature_ind = x_ind_ptr[j]
                     val = x_data_ptr[j]
-                    sum_gradient[idx] += gradient_change * val
+                    for class_ind in range(n_classes):
+                        sum_gradient[feature_ind * n_classes + class_ind] += \
+                            val * (gradient[class_ind] -
+                                   gradient_memory[sample_ind * n_classes +
+                                                   class_ind])
 
+                # fit the intercept
                 if fit_intercept:
-                    intercept_sum_gradient += gradient_change
-                    intercept -= (step_size * intercept_sum_gradient
-                                  / num_seen * intercept_decay)
+                    for class_ind in range(n_classes):
+                        intercept_sum_gradient[class_ind] += \
+                            (gradient[class_ind] -
+                             gradient_memory[sample_ind * n_classes +
+                                             class_ind])
+                        intercept[class_ind] -= \
+                            (step_size * intercept_sum_gradient[class_ind] /
+                             num_seen * intercept_decay)
+
+                        # check to see that the intercept is not inf or NaN
+                        if not skl_isfinite(intercept[class_ind]):
+                            with gil:
+                                raise_infinite_error(n_iter)
 
                 # update the gradient memory for this sample
-                gradient_memory[current_index] = gradient
+                idx = sample_ind * n_classes - 1
```

Comment:
You could cache this for the above two loops as well, for the one across the non zero features and the intercept loop across n_classes